predicatefs.go

```diff
@@ -0,0 +1,219 @@
+}
+
+func (p *FilePredicateFs) Chown(path string, uid, gid int) error {
+	if err := p.dirOrValidPath(path); err != nil {
+		return err
+	}
+	return p.source.Chown(path, uid, gid)
+}
+
+func (p *FilePredicateFs) Name() string {
+	return "FilePredicateFs"
+}
+
+func (p *FilePredicateFs) Stat(path string) (os.FileInfo, error) {
+	if err := p.dirOrValidPath(path); err != nil {
+		return nil, err
+	}
+	return p.source.Stat(path)
+}
+
+func (p *FilePredicateFs) Rename(oldname, newname string) error {
+	dir, err := IsDir(p.source, oldname)
+	if err != nil {
+		return err
+	}
+	if dir {
+		return nil
+	}
+	if err := p.validate(oldname); err != nil {
+		return err
+	}
+	if err := p.validate(newname); err != nil {
+		return err
+	}
+	return p.source.Rename(oldname, newname)
+}
+
+func (p *FilePredicateFs) RemoveAll(path string) error {
+	dir, err := IsDir(p.source, path)
+	if err != nil {
+		return err
+	}
+	if !dir {
+		if err := p.validate(path); err != nil {
+			return err
+		}
+	}
+	return p.source.RemoveAll(path)
+}
+
+func (p *FilePredicateFs) Remove(path string) error {
+	if err := p.dirOrValidPath(path); err != nil {
+		return err
+	}
+	return p.source.Remove(path)
+}
+
+func (p *FilePredicateFs) OpenFile(path string, flag int, perm os.FileMode) (File, error) {
+	if err := p.dirOrValidPath(path); err != nil {
+		return nil, err
+	}
+	return p.source.OpenFile(path, flag, perm)
+}
+
+func (p *FilePredicateFs) Open(path string) (File, error) {
+	dir, err := IsDir(p.source, path)
+	if err != nil {
+		return nil, err
+	}
+	if !dir {
+		if err := p.validate(path); err != nil {
+			return nil, err
+		}
+	}
+	f, err := p.source.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	return &PredicateFile{f: f, pred: p.pred}, nil
+}
+
+func (p *FilePredicateFs) Mkdir(n string, path os.FileMode) error {
+	return p.source.Mkdir(n, path)
+}
+
+func (p *FilePredicateFs) MkdirAll(n string, path os.FileMode) error {
+	return p.source.MkdirAll(n, path)
+}
+
+func (p *FilePredicateFs) Create(path string) (File, error) {
+	if err := p.validate(path); err != nil {
+		return nil, err
+	}
+	return p.source.Create(path)
+}
+
+func (f *PredicateFile) Close() error {
+	return f.f.Close()
+}
+
+func (f *PredicateFile) Read(s []byte) (int, error) {
+	return f.f.Read(s)
+}
+
+func (f *PredicateFile) ReadAt(s []byte, o int64) (int, error) {
+	return f.f.ReadAt(s, o)
+}
+
+func (f *PredicateFile) Seek(o int64, w int) (int64, error) {
+	return f.f.Seek(o, w)
+}
+
+func (f *PredicateFile) Write(s []byte) (int, error) {
+	return f.f.Write(s)
+}
+
+func (f *PredicateFile) WriteAt(s []byte, o int64) (int, error) {
+	return f.f.WriteAt(s, o)
+}
+
+func (f *PredicateFile) Name() string {
+	return f.f.Name()
+}
+
+func (f *PredicateFile) Readdir(c int) (fi []os.FileInfo, err error) {
+	var pfi []os.FileInfo
+	pfi, err = f.f.Readdir(c)
+	if err != nil {
+		return nil, err
+	}
+	for _, i := range pfi {
+		if i.IsDir() || f.pred(filepath.Join(f.f.Name(), i.Name())) {
```

Comment:
@satotake I'm not sure what you mean by "skip directories"; to me it looks like every directory is included.

I would have found it much more useful if the above was replaced with:

```go
if f.pred(i.IsDir(), filepath.Join(f.f.Name(), i.Name())) {
```

Reply:
One might as well just pass the file info object instead of a bool which can then be manually checked whether the path is a directory or a file or a symlink, etc.

Reply:
That would break if you want to use this with `fs.ReadDirFile` (see my recent commit to the main branch).

Reply:
@bep I got your point.
`isDir` will be nice to ignore some dirs without scanning their children, for instance.
Fixed.